File: basic/lua/bfs/bfs.lua
#!/usr/local/bin/lua

function split(str,delimiter)
  if str==nil or str=="" or delimiter==nil
  then
    return nil
  end
  local i=1
  local j=0
  local k=0
  local result={}
  while true do
    j,k=string.find(str,delimiter,i)
    if j == nil
    then
      break
    end
    -- print(string.sub(str,i,j))
    table.insert(result,string.sub(str,i,j))
    i=k+1
  end
  -- print(i)
  table.insert(result,string.sub(str,i))
  return result
end
-- split("1 1"," ")

mazeArr={}
bookArr={}
startCoord={}
targetCoord={}
head=1
tail=1
que={}
nextStep={{-1,0},{0,1},{1,0},{0,-1}}

function bfs()
  currentCoord = {que[head][1],que[head][2]}
  flag=false
  while head<tail do
    print("head"..head)
    print("tail"..tail)
    for i=1,4 do

      currentCoord[1]=que[head][1]+nextStep[i][1]
      currentCoord[2]=que[head][2]+nextStep[i][2]
      if currentCoord[1]<= 0 or currentCoord[1]>#mazeArr or currentCoord[2]<= 0 or currentCoord[2]>#mazeArr[1] then

      else
        print("c1:"..currentCoord[1])
        print("c2:"..currentCoord[2])
        print(mazeArr[currentCoord[1]][currentCoord[2]])
        print(bookArr[currentCoord[1]][currentCoord[2]])

        if mazeArr[currentCoord[1]][currentCoord[2]]-0==0 and bookArr[currentCoord[1]][currentCoord[2]] == 0 then
          -- 不是障碍物，并且没有走过
          bookArr[currentCoord[1]][currentCoord[2]]=1
          table.insert(que,{currentCoord[1],currentCoord[2],que[head][3]+1})
          tail=tail+1
        end
        if currentCoord[1]==targetCoord[1] and currentCoord[2]==targetCoord[2] then
            flag=true
            break
        end
      end
    end
    if flag then
        break
    end
    head=head+1
  end
  print('最少的步数为:')
  print(que[#que][3])
end



print("输入迷宫,0表示空,1表示有障碍物,以空行结束")
while true do
  temp=io.read("*line")
  if temp == "" then
    break
  end
  table.insert(mazeArr,split(temp," "))
end
print("输入起始点坐标:")
startCoord=split(io.read("*line")," ")
for i,v in ipairs(startCoord) do
	startCoord[i] = tonumber(v)
end
print("输入目标点坐标:")
targetCoord=split(io.read("*line")," ")
for i,v in ipairs(targetCoord) do
	targetCoord[i] = tonumber(v)
end
for i,v in ipairs(mazeArr) do
  tempArr={}
	for j,k in ipairs(v) do
    table.insert(tempArr,0)
  end
  table.insert(bookArr,tempArr)
end
table.insert(que,{startCoord[1],startCoord[2],0})
tail = tail+1

bfs()

File: basic/perl/bfs/bfs.pl
#!/usr/bin/perl

@mazeArr=();
@bookArr=();
@startCoord=();
@targetCoord=();

# 走下一步的坐标
@nextStep=((-1,0),(0,1),(1,0),(0,-1));
# 最小步数
$minStep=0;
@que=();
$head=0;
$tail=0;

print("输入迷宫,0表示空,1表示有障碍物,以空行结束\n");
while(true){
  $line = <STDIN>;
  if($line eq "\n") {
    last;
  } else {
    @temp =  split(" ",$line);
    push(@mazeArr,[@temp]);
  }
  print @mazeArr

}
print "输入起始坐标:";
$line = <STDIN>;
@startCoord=split(" ",$line);
# print @startCoord;
print "输入目标坐标:";
$line = <STDIN>;
@targetCoord=split(" ",$line);
# print @targetCoord;


for($i=0;$i<=$#mazeArr;$i+=1) {
  print $#mazeArr;
  print "\n";
  foreach $tempVal ($mazeArr[i]) {
   push(@{$bookArr[$i]},"0");
  }
  print @{$bookArr[$i]}

}

# foreach $tempArr (@mazeArr) {
#   @tempArr2=();
#   foreach $tempVal ($tempArr) {
#     # push(@tempArr2,"0");
#     push(@{$bookArr[]})
#   }
#
#   # push(@bookArr,[@tempArr2]);
# }

print @bookArr;
print @bookArr[1]->[1];

bfs();

sub bfs {

  $flag=false;
  while($head<$tail) {
    for($i=0;i<4;$i+=1){
      @currentCoord=($que[$head][0]+$nextStep[i][0],$que[$head][1]+$nextStep[i][1]);
          # if ($currentCoord[0]<0 or $currentCoord[0]> $#mazeArr or $currentCoord[1]<0 or $currentCoord[1]>scalar($mazeArr[0]) {
          #
          # }
    }
  }

}
